### ISSUE TITLE
Slices have parent content specific type names

### ISSUE THREAD

**Issue by pr1ntr** [angeloashmore/gatsby-source-prismic#122]
I use the same slices across different pages, however I am noticing that if i use slice: **GridModule** on **PrismicLandingPage** and **PrismicRestaurantLandingPage** that is has these type names respectively:

**PrismicLandingPageBodyGridModule**
**PrismicRestaurantLandingPageBodyGridModule**

Since a Prismic repo uses all slices generically across all content types where they allowed in the slice zone I don't think that the plugin should create duplicate types with the same properties.

I would like to be able to just define a fragment:

```
fragment GridModuleContent on PrismicSliceGridModule { 


}
```

instead of

```
fragment GridModuleContent on PrismicRestaurantLandingPageBodyGridModule {
  
}

```

**Comment by angeloashmore** [comment]
Hi @pr1ntr, this is intentional since Prismic does not allow true generic slices. The slice library feature allows you to effectively copy/paste slices across content types, but nothing stops you from altering it after the initial paste.

Thus, it is required to namespace them since `PrismicLandingPageBodyGridModule` could have slightly different fields from `PrismicRestaurantLandingPageBodyGridModule` despite the name sharing.

If you have any ideas of how this could be improved, please share! Ideally Prismic would allow true generic/referential slices, but that is not how it is currently implemented.

**Comment by pr1ntr** [comment]
Yeah I totally realize this is an aspect of Prismic. So here is my thought:

While the Slices schemas in Prismic are not generic, they behave as if they were.
For instances the key name under config.choices in slices is always the same if you use the same slice:
```
 "config": {
        "choices": {
          "grid_module": {
````

in this case its always grid_module. So far I am assuming this will be the case with all slices. Personally I think it would be safe to rework the node creation to pull out body.type === 'slices' and generate a coherent normalized schema to match the UX/UI of Prismic. 

That being said, I can understand reluctance because now we are making assumptions that are not inherent to the system's design but to the user experience. Personally, I think it would improve DX if we did this. Though I would like to get some differing opinions.

**Comment by Ehesp** [comment]
This is something I've also been wondering.

Is it possible to maybe merge all of the similar fields together? E.g.

Slice FooBar has the field (on page A):

- title

Slice FooBar has the fields (on page B):

- title
- description

Even though slice has been modified on page B, it's still the slice FooBar. So when querying it on page A, the description field will just return null? Kind of merging all of the uses of FooBar together?

**Comment by pr1ntr** [comment]
This is basically what I am thinking this lib should do. The only issue I can see is that since Prismic does not have any generic types this behavior could cause unexpected side effects. If it were me, I would just do it anyways because the side effects it would cause would be because poor schema design on the cms authors part.

In order to get around this limitation I had to just stop using pagequeries. I write my query content kinda like this:

```javascript
const buildSliceQuery = (prefix) => (result, value, key) => result + `
    ... on ${prefix + key} {
      ${value}
    }
  `

const RestaurantLandingPage = `
  ... on PrismicRestaurantLandingPage {
    ${pageProps}
    data {
      title
      body {
        ${reduce(Slices, buildSliceQuery(PrefixRestaurantLandingPage), '')}
      }
    }
  }
`
```

Where I pass the prefix to the predicate function and let it concat it to the slice name. All of this happens in gatsby-node and is passed into pageContext. 

This is the second time I had to just throw out pagequeries because of how ugly and spaghetti like they become unless you have a perfectly design graphql schema that is almost completely generic.

**Comment by angeloashmore** [comment]
My initial reaction is that it seems a bit too volatile and magical. These auto-generated fragments would change automatically as the custom type schemas change. The changes may seem to come out of nowhere without deep understanding of how the fragments are generated.

It would either freeze parts of your schemas or require rewriting your queries when changes occur.

@pr1ntr Out of curiosity, where are you using those strings? Gatsby's `graphql` tagged template literal does not allow interpolation since it must be statically available at build time.

To be clear, I totally understand the frustration. I've worked around this issue by using a generic "Page" custom type that includes all possible slices. This gives the advantage of using any slice on any page while removing the need to duplicate slice configs. There are trade-offs here, of course, and this strategy may not apply to all use cases, but it's worked well for me.

**Comment by pr1ntr** [comment]
@angeloashmore I am running these queries in `gatsby-node.js::exports.createPages`


`SITE_PAGES` is the full query for the all the pages.

```javascript
exports.createPages = async ({ graphql, actions }) => {
  const { createPage } = actions
  try {
    const { data } = await wrapper(
      graphql(SITE_PAGES)
    )

    const merged = mergePageData(data)

    const [ currentlyActive, futureDeployments ] = parsePublishedList(merged)

    currentlyActive.forEach((entry) => {

      // The uid you assigned in Prismic is the slug!

      let path = entry.data.path

      if (entry.uid) {
        path+=`/${entry.uid}`
      }

      createPage({
        path,
        component: TEMPLATES[entry.type] || DEFAULT_TEMPLATE,
        context: {
          // Pass the unique ID (uid) through context so the template can filter by it
          id: entry.id,
          pagePath: path,
          staticData: entry,
        },
      })
    
    })

    if (process.env.NODE_ENV === 'development') {
      createPage({
        path: '/preview',
        component: PREVIEW_TEMPLATE,
        context: {
          pages: merged,
        },
      })
    }
  } catch (error) {
    console.error('error', error)
  }
}
```

**Comment by madeleineostoja** [comment]
> I've worked around this issue by using a generic "Page" custom type that includes all possible slices. This gives the advantage of using any slice on any page while removing the need to duplicate slice configs. There are trade-offs here, of course, and this strategy may not apply to all use cases, but it's worked well for me.

@angeloashmore I realise this issue is really old, but I've also just run into this limitation on a client project with extensive slice reuse — could you share a little more about how your setup works? How does having a single type with all slices in it help when querying other documents? Creating fragments for that type still wouldn't work in a different type with copied slices right?

My particular use case is a slice library where every single slice is reused in about 4 different content models. Basically a website builder but with Prismic lol.

**Comment by angeloashmore** [comment]
Hey @madeleineostoja, no worries about this being an old issue. It's still relevant!

What I meant in that comment is to use a single document type for every public page on a site. So rather than having four different document types with duplicated slices, you have just one with **all** slices. Any non-public-page content, such as a shop Product or Navigation links, is managed in its own document type and can be referenced by a "Public Page" slice.

I also treat Prismic as a website builder. Every site page is just a list of slices. Even headers and footers are slices (although these don't get added to Prismic, just treated as such in code).

For large sites with a lot of visual variation, this strategy may not be possible. For example a site with 40 slice types would be extremely difficult to manage for editors. I've reached this point and it's not pretty…

**Comment by angeloashmore** [comment]
I haven't tried this, but you could possibly use Gatsby's `createSchemaCustomization` Node API to manually setup a shared interface. If you are 100% sure the fields between a subset of slices are 100% the same, at least type-wise, you should be able to create a GraphQL interface with those slices and tell Gatsby that those slices implement the interface.

You could then use a fragment on those slices.

Let me know if you need me to write out an example if that doesn't make sense.

**Comment by madeleineostoja** [comment]
Ahhh okay, gotcha! That's a good idea, I might try and convince my client to rebuild the site a single mega document type, with everything as a slice. It's a bit messy and I don't love it, I really wish Prismic just had real reusable slices. The slice library is a bit of a joke and is asking for things to fall out of sync and break.

**Comment by angeloashmore** [comment]
I'm cleaning out old issues and closing this due to its age. Sharing Slice types will be possible with Shared Slices once Prismic's [Slice Machine](https://www.slicemachine.dev/) works with Gatsby (in development!).
